File: 2787. Ways to Express an Integer as Sum of Powers/Solution.cpp
class Solution {
public:
    int numberOfWays(int n, int x) {
        vector<int> powers;
        int mod = 1000000007;
        for (int i = 1; ; i++) {
            int curr = i;
            for (int j = 1; j < x; j++) curr *= i;
            if (curr > n) break;
            powers.push_back(curr);
        }
        int p = powers.size();
        vector<int> dp(n + 1, 0);
        dp[0] = 1;
        for (int i : powers) {
            for (int j = n; j >= 0; j--) {
                if (j >= i) dp[j] = (dp[j] + dp[j - i]) % mod;
            }
        }
        return dp[n];
    }
};

File: 2787. Ways to Express an Integer as Sum of Powers/Solution.py
class Solution:
    def numberOfWays(self, n: int, x: int) -> int:
        powers = []
        curr, mod = 1, 1000000007
        while curr ** x <= n:
            powers.append(curr ** x)
            curr += 1
        p = len(powers)
        dp = [0 for i in range(n + 1)]
        dp[0] = 1
        for i in powers:
            for j in range(n, -1, -1):
                if j >= i:
                    dp[j] = (dp[j] + dp[j - i]) % mod
        return dp[-1]
